csharp/src/LibUSearch/USearchIndex.cs

```diff
@@ -0,0 +1,262 @@
+    {
+        this.CheckIncreaseCapacity(1);
+        usearch_add(this._index, key, vector, usearch_scalar_kind_t.usearch_scalar_f32_k, out IntPtr error);
+        HandleError(error);
+    }
+
+    public void Add(ulong key, double[] vector)
+    {
+        this.CheckIncreaseCapacity(1);
+        usearch_add(this._index, key, vector, usearch_scalar_kind_t.usearch_scalar_f64_k, out IntPtr error);
+        HandleError(error);
+    }
+
+    public void AddMany(ulong[] keys, float[][] vectors)
+    {
+        this.CheckIncreaseCapacity((ulong)vectors.Length);
+        for (int i = 0; i < vectors.Length; i++)
+        {
+            usearch_add(this._index, keys[i], vectors[i], usearch_scalar_kind_t.usearch_scalar_f32_k, out IntPtr error);
+            HandleError(error);
+        }
+    }
+
+    public void AddMany(ulong[] keys, double[][] vectors)
+    {
+        this.CheckIncreaseCapacity((ulong)vectors.Length);
+        for (int i = 0; i < vectors.Length; i++)
+        {
+            usearch_add(this._index, keys[i], vectors[i], usearch_scalar_kind_t.usearch_scalar_f64_k, out IntPtr error);
+            HandleError(error);
+        }
+    }
+
+    public bool Get(ulong key, out float[] vector)
+    {
+        vector = new float[this._cached_dimensions];
+        bool success = usearch_get(this._index, key, vector, usearch_scalar_kind_t.usearch_scalar_f32_k, out IntPtr error);
+        HandleError(error);
+        if (!success)
+        {
+            vector = null;
+        }
+
+        return success;
+    }
+
+    public bool Get(ulong key, out double[] vector)
+    {
+        vector = new double[this._cached_dimensions];
+        bool success = usearch_get(this._index, key, vector, usearch_scalar_kind_t.usearch_scalar_f64_k, out IntPtr error);
+        HandleError(error);
+        if (!success)
+        {
+            vector = null;
+        }
+
+        return success;
+    }
+
+    private ulong Search<T>(T[] queryVector, ulong resultsLimit, out Dictionary<ulong, float> foundKeyDistances, usearch_scalar_kind_t scalarKind)
+    {
+        ulong[] keys = new ulong[resultsLimit];
+        float[] distances = new float[resultsLimit];
+
+        GCHandle handle = GCHandle.Alloc(queryVector, GCHandleType.Pinned);
+        ulong matches = 0;
+        try
+        {
+            IntPtr queryVectorPtr = handle.AddrOfPinnedObject();
+            matches = (ulong)usearch_search(this._index, queryVectorPtr, scalarKind, (UIntPtr)resultsLimit, keys, distances, out IntPtr error);
+            HandleError(error);
+        }
+        finally
+        {
+            handle.Free();
+        }
+        foundKeyDistances = new Dictionary<ulong, float>();
+        for (ulong i = 0; i < matches; i++)
+        {
+            foundKeyDistances.Add(keys[i], distances[i]);
+        }
+
+        return matches;
+    }
+
+    public ulong Search(float[] queryVector, ulong resultsLimit, out Dictionary<ulong, float> foundKeyDistances)
```

Comment:
Is it a good idea to export results into a dictionary? The order of entries will be lost, right?

Reply:
No way it is a good! Thanks for pointing that out. I recently added a commit where I fixed that.